aesd-char-driver/main.c

```diff
@@ -0,0 +1,286 @@
+    //Check for the results
+    if(!entry)
+    {
+        PDEBUG("Access to aesdchar is returning without contents read\n");
+        mutex_unlock(&aesd_device.lock);
+        *f_pos = 0;
+        retval = 0;
+        return retval;
+    }
+    //Copy from entry offset, "count" or size
+    if(entry_offset + count >= entry->size)
+    {
+        //Copy only until the size
+        if(copy_to_user(buf, &entry->buffptr[entry_offset],  entry->size - entry_offset) != 0)
+        {
+            PDEBUG("Could not copy memory to the user\n");
+            mutex_unlock(&aesd_device.lock);
+            retval = -EINVAL;
+            return retval;
+        }
+        *f_pos = *f_pos + entry->size;
+        retval = entry->size - entry_offset;
+    }
+    else
+    {
+        //Copy until the count
+        //Copy only until the size
+        if(copy_to_user(buf, &entry->buffptr[entry_offset],  count) != 0)
+        {
+            PDEBUG("Could not copy memory to the user\n");
+            mutex_unlock(&aesd_device.lock);
+            retval = -EINVAL;
+            return retval;
+        }
+        *f_pos = *f_pos + count;
+        retval = count;
+    }
+
+    //Release mutex
+    mutex_unlock(&aesd_device.lock);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    const char *to_be_freed = NULL;
+    PDEBUG("write %zu bytes with offset %lld\n",count,*f_pos);
+    
+
+    //Check filp and buf validity
+    if(!filp || !buf)
+    {
+        PDEBUG("Could not reallocate a temporary command\n");
+        retval = -EINVAL;
+        return retval;
+    }
+
+    //Check that the count is not 0
+    if(!count)
+    {
+        PDEBUG("Attempt to write 0 bytes\n");
+        retval = 0;
+        return retval;
+    }
+
+    //Set mutex
+    mutex_lock(&aesd_device.lock);
+
+    //Check if there is a partial write active
+    if(aesd_device.partial_size)
+    {
+        aesd_device.partial_content = krealloc(aesd_device.partial_content, 
+                                              sizeof(char)*(aesd_device.partial_size + count), GFP_KERNEL); 
+        if(!aesd_device.partial_content)
+        {
+            PDEBUG("Could not reallocate a temporary command\n");
+            mutex_unlock(&aesd_device.lock);
+            retval = -ENOMEM;
+            return retval;
+        }
+    }
+    //If not, allocate temporary space
+    else
+    {
+        aesd_device.partial_content = kmalloc(sizeof(char)*count, GFP_KERNEL);
+        if(!aesd_device.partial_content)
+        {
+            PDEBUG("Could not allocate a temporary command\n");
+            mutex_unlock(&aesd_device.lock);
+            retval = -ENOMEM;
+            return retval;
+        }
+    }
+    
+    //Check every value in the entry and commit if '\n' is found
+    while(count)
+    {
+        get_user(aesd_device.partial_content[aesd_device.partial_size], &buf[retval]);
```

Comment:
@jordicrosmompart I think you could simplify the logic here with `copy_from_user`